### PR TITLE
fixed OnUnitEnterLeave

### PR DESCRIPTION
**Pull request by IgorSamurovic**
An update due to bugs discussed in #13 (both should be solved by this).

### CODE REVIEW

File: wurst/event/OnUnitEnterLeave.wurst [hunk 1]
```diff
@@ -31,22 +32,46 @@ public function onLeave(code c)
 	eventTrigger.addAction(c)
 
 /* === Internals === */
+constant UNITS_PER_CALL = 50
+constant ABILITY_ID = compiletime(ABIL_ID_GEN.next())
 
 constant eventTrigger = CreateTrigger()
 unit tempUnit = null
+group preplacedUnits
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: -> `let preplacedUnits = CreateGroup()`

File: wurst/event/OnUnitEnterLeave.wurst [hunk 2]
```diff
@@ -55,12 +80,10 @@ init
 			tempUnit = null
 	end)
 
-	GroupEnumUnitsInRect(ENUM_GROUP, boundRect, null)
-	for u from ENUM_GROUP
-		if u == null
-			break
-		u..addAbility(ABILITY_ID)..makeAbilityPermanent(ABILITY_ID, true)
-		eventTrigger.evaluate()
+	preplacedUnits = CreateGroup()
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: Remove

File: wurst/event/OnUnitEnterLeave.wurst [hunk 3]
```diff
@@ -31,22 +32,46 @@ public function onLeave(code c)
...
 unit tempUnit = null
+group preplacedUnits
+constant trigger preplacedUnitsCall = CreateTrigger()..addAction(function initPreplacedUnits)
 
-constant ABILITY_ID = compiletime(ABIL_ID_GEN.next())
+function prepareUnit(unit u)
+	u..addAbility(ABILITY_ID)..makeAbilityPermanent(ABILITY_ID, true)
+
+function initPreplacedUnits()
+	for count=1 to UNITS_PER_CALL
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: Can you invert the order of control on this for groups vs counting?

As in:

```
var count = UNITS_PER_CALL

for u from preplacedUnits
    prepareUnit(u)
    eventTrigger.evaluate()
    count--
    if count <= 0
        break
```

File: wurst/event/OnUnitEnterLeave.wurst [hunk 4]
```diff
@@ -31,22 +32,46 @@ public function onLeave(code c)
...
+		let u = FirstOfGroup(preplacedUnits)
+		if u == null
+			break
+		else
+			preplacedUnits.removeUnit(u) 
+			prepareUnit(u)
+			eventTrigger.evaluate()
+
+	if FirstOfGroup(preplacedUnits) != null
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: use `preplacedUnits.hasNext()`

File: wurst/event/OnUnitEnterLeave.wurst [hunk 5]
```diff
@@ -55,12 +80,10 @@ init
...
 	end)
 
-	GroupEnumUnitsInRect(ENUM_GROUP, boundRect, null)
-	for u from ENUM_GROUP
-		if u == null
-			break
-		u..addAbility(ABILITY_ID)..makeAbilityPermanent(ABILITY_ID, true)
-		eventTrigger.evaluate()
+	preplacedUnits = CreateGroup()
+	GroupEnumUnitsInRect(preplacedUnits, boundRect, null)
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: Can you add, and then use: `group.enumUnitsInRect(rect r)`?

**Comment by Cokemonkey11** [comment]
Can you resolve the notes I made?

Overall, looks okay. I'd probably pay a little overhead to make it more forward declarative intead of recursive. Something like this:

```
while grp.hasNext()
    evaluate processGroupBatch(grp) // processes at most 50 units in a separate jthread
```

File: wurst/event/OnUnitEnterLeave.wurst [hunk 6]
```diff
@@ -36,24 +36,24 @@ constant ABILITY_ID = compiletime(ABIL_ID_GEN.next())
...
 			break
-		else
-			preplacedUnits.removeUnit(u) 
-			prepareUnit(u)
-			eventTrigger.evaluate()
 
-	if FirstOfGroup(preplacedUnits) != null
-		preplacedUnitsCall.execute()
+	if preplacedUnits.hasNext()
+		preplacedUnitsCall.execute() // processes at most UNITS_PER_CALL units per thread
```
Comment (Cokemonkey11) [inline wurst/event/OnUnitEnterLeave.wurst]: I also used that word "thread", but it's not really correct.

Dracolich showed that this op-limit is a single value incrementing, so maybe `processes at most UNITS_PER_CALL units before resetting the op-limit.`

**Comment by Cokemonkey11** [comment]
Looks good enough to me now. @Frotty ?

**Comment by Cokemonkey11** [comment]
@IgorSamurovic no UTs? Can you confirm this works for < 50 and > 150 units ?

**Comment by IgorSamurovic** [comment]
I have tested this for 30 and 300 units and it works without any problems. I haven't figured out how UTs work in wurst yet.

**Comment by IgorSamurovic** [comment]
Tested again with ForGroup, no problems.

File: wurst/event/OnUnitEnterLeave.wurst [hunk 7]
```diff
@@ -37,26 +37,10 @@ constant ABILITY_ID = compiletime(ABIL_ID_GEN.next())
 constant eventTrigger = CreateTrigger()
 unit tempUnit = null
 let preplacedUnits = CreateGroup()
-constant trigger preplacedUnitsCall = CreateTrigger()..addAction(function initPreplacedUnits)
 
 function prepareUnit(unit u)
 	u..addAbility(ABILITY_ID)..makeAbilityPermanent(ABILITY_ID, true)
```
Comment (Frotty) [inline wurst/event/OnUnitEnterLeave.wurst]: move `eventTrigger.evaluate()`  here

Comment (IgorSamurovic) [reply wurst/event/OnUnitEnterLeave.wurst]: What do you mean?

Comment (Frotty) [reply wurst/event/OnUnitEnterLeave.wurst]: eventTrigger.evaluate() is called after each time you use prepareUnit.
So move it inside the function.

Comment (IgorSamurovic) [reply wurst/event/OnUnitEnterLeave.wurst]: Oh yeah you're right. Done.

**Comment by Frotty** [comment]
Nice, thanks a lot! ♉️

**Pull request merged**